### PR TITLE
csm-portal: add exact-match caseNumber/wso2CaseId case search filters

### PR DESCRIPTION
**Pull request by rksk**
## Purpose
The CSM portal's global case search was slow (multiple seconds in production for a single case-number lookup). Case number and WSO2 case id were only reachable through the free-text `searchQuery` filter, which the backing ServiceNow data source evaluates as an unindexed scan across four columns (case number, WSO2 case id, subject, description) -- expensive regardless of how precise the typed value is, and worse for the CSM portal than for the customer portal because the CSM search has no per-account scoping to narrow the candidate set first.

## Goals
Add `caseNumber` and `wso2CaseId` as first-class, exact-match case-search filters (mirroring how every other simple case attribute -- state, severity, project, parentId, etc. -- already works), and route the CSM portal's global quick-search palette to use them automatically when the typed text looks like a case number or a WSO2 case id, instead of always going through the free-text scan.

## Approach
Three layers, bottom-up:
- **entity-service**: added `caseNumber`/`wso2CaseId` to the case-search field-filter allowlist (`eq`-only, single value, same pattern as the existing `parentId` filter), threaded end-to-end into the request sent to the backing data source. Documented in `openapi.yaml`.
- **csm-portal backend**: this service's case-search path is a pure byte passthrough (no typed request struct), so no Go code change was needed -- only its own independently-maintained `openapi.yaml` was updated to keep the documented contract in sync.
- **csm-portal webapp**: the global quick-search palette (`QuickNav`) now classifies the typed query -- a `CS` + 7-digit pattern routes to the `caseNumber` filter, a `<prefix>-<1-4 digits>` pattern routes to `wso2CaseId`, anything else keeps using the existing free-text search unchanged. When a scoped exact-match search runs, an inline banner tells the user the search was narrowed to that attribute, with a "Search all fields" action to re-run the same query as free text -- narrowing is never silent, and widening is always the user's explicit choice.

A corresponding change to the entity-service's backing data source (additive, gated behind the two new optional filter keys, verified live against a non-production environment) has already landed separately.

## User stories
As a CS engineer using the CSM portal's quick-search, when I type a known case number or WSO2 case id, I want the matching case to come back near-instantly instead of waiting several seconds, since this is the most common thing I search for during a support call.

## Release note
Case search now supports exact-match filtering by case number (`caseNumber`) and WSO2 case id (`wso2CaseId`), in addition to the existing free-text search. The CSM portal's global quick-search automatically uses the faster exact-match path when the typed text looks like a case number or WSO2 case id, with an option to widen back to a full-text search.

## Documentation
N/A -- internal API filter addition, no external-facing documentation to update.

## Training
N/A -- no training content impact.

## Certification
N/A -- no certification exam impact.

## Marketing
N/A -- internal performance fix, no marketing content.

## Automation tests
 - Unit tests
   > entity-service: added `ParseCaseFieldFilters` parse/rejection tests for both new fields (`entity-service/internal/service/case_filters_test.go`), following the existing `parentId` test pattern. Full `go test ./...` passes for both `entity-service` and `apps/csm-portal/backend`.
 - Integration tests
   > csm-portal webapp: added routing tests for `classifyQuickCaseQuery` (both regexes' edge cases) and `useQuickCaseSearch` (all three routing branches, including the `forceFreeText` widen override), plus `QuickNav` tests for the new "exact match" banner and its widen action. Full `vitest` suite passes with no regressions (986 passed / 8 pre-existing unrelated failures, confirmed present on `main` before this change).

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
 - Ran FindSecurityBugs plugin and verified report? N/A -- Go and TypeScript changes only, no Java in this PR.
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

## Samples
N/A -- no sample apps affected.

## Related PRs
A corresponding backing-data-source contract change has already landed separately (tracked internally, not linked here as it lives in a private repo).

## Migrations (if applicable)
N/A -- additive filter fields, no data migration.

## Test environment
Verified via `go build`/`go vet`/`go test` (entity-service, csm-portal backend) and `tsc -b`/`vitest` (csm-portal webapp) locally. No browser-specific behavior introduced.

## Learning
N/A.

## Summary by CodeRabbit

- **New Features**
  - Quick navigation now recognizes exact case, change request, incident, and problem numbers.
  - Case searches also support exact internal ID matching.
  - Exact matches search identifiers directly for more precise results.
  - Added an option to broaden exact-match searches to free text.
  - Added clearer search messages, exact-match indicators, widening actions, and format guidance.
  - Improved deep-link handling, loading states, and empty-state behavior.

### CODE REVIEW

File: apps/csm-portal/webapp/src/App.tsx [hunk 1]
```diff
@@ -152,13 +160,28 @@ const CsmAnnouncementsPage = lazy(
...
+ *
+ * A pure read of sessionStorage/search params — AuthGuard and QuickNav own
+ * clearing their respective keys.
  */
 function RootLanding(): JSX.Element | null {
   const pending = sessionStorage.getItem(POST_LOGIN_REDIRECT_KEY);
-  return pending ? null : <Navigate to="/dashboard" replace />;
+  const [searchParams] = useSearchParams();
+  const hasDeepLinkSearch = searchParams.has("goto") || searchParams.has("q");
+  return pending || hasDeepLinkSearch ? null : <Navigate to="/dashboard" replace />;
```
[the remark this answers is not in the document]
Comment (rksk) [reply apps/csm-portal/webapp/src/App.tsx:191]: Confirmed and fixed in `e8b3abf1c` — `hasDeepLinkSearch` now requires a non-blank trimmed value on `goto`/`q`, matching the suggested fix. Didn't need the additional QuickNav-side change: its own initial-params effect already no-ops on an empty value (`if (!q && !goto) return`), so once RootLanding stops deferring, nothing was left to clear.

File: apps/csm-portal/webapp/src/features/csm-cases/api/useQuickCaseSearch.test.tsx [hunk 2]
```diff
@@ -27,7 +27,10 @@ vi.mock("@api/backend/client", () => ({
   useBackendApi: () => ({ post: postMock }),
 }));
 
-import { useQuickCaseSearch } from "@features/csm-cases/api/useQuickCaseSearch";
+import {
+  classifyQuickCaseQuery,
+  useQuickCaseSearch,
+} from "@features/csm-cases/api/useQuickCaseSearch";
```
[the remark this answers is not in the document]
Comment (rksk) [reply apps/csm-portal/webapp/src/features/csm-cases/api/useQuickCaseSearch.test.tsx:33]: Checked: no-op, both test files already pass (14/14 across the two). Neither `useQuickCaseSearch.ts`/`useQuickChangeRequestSearch.ts` nor their transitive imports touch `@config/apiConfig` — the only path that would is `@api/backend/client`, which both files already mock wholesale (`vi.mock("@api/backend/client", () => ({ useBackendApi: () => ({ post: postMock }) }))`), so the real client module (and its `apiConfig` read) never executes. Grepped for `apiConfig` across both hooks' full import chains (mappers, apiConstants, caseType utils) — zero hits. Leaving as-is rather than adding an unused mock.

File: apps/csm-portal/webapp/src/features/csm-recent/components/QuickNav.tsx [hunk 3]
```diff
@@ -141,18 +280,46 @@ export default function QuickNav(): JSX.Element | null {
...
+  // `showExactMatchBanner` by construction: this requires
+  // `exactMatchKind === null`, that requires `exactMatchKind !== null`.
+  // Deliberately NOT gated on `anySearchFetching`/`resultsSettling`: it
+  // needs to stay visible through a settled zero-hit result too (e.g.
+  // typing "cs123" — doesn't match the exact-match shape, free-text search
+  // finds nothing), not just disappear the moment the search finishes,
+  // which would leave "No matches." with no explanation of why a plain-
+  // looking query didn't get the fast exact-match path.
+  const showNoPatternHint =
+    !forceFreeText && exactMatchKind === null && trimmedQuery.length > 0;
```
[the remark this answers is not in the document]
Comment (rksk) [reply apps/csm-portal/webapp/src/features/csm-recent/components/QuickNav.tsx]: Fixed in `e8b3abf1c` — `showNoPatternHint` now also requires `trimmedQuery.length >= QUICK_CASE_MIN_QUERY_LEN` (all four hooks' min length happens to be 2, so using that constant as the representative one).

File: entity-service/internal/domain/entity.go [hunk 4]
```diff
@@ -2207,6 +2217,11 @@ type SearchChangeRequestsFilters struct {
 	Impacts         []ChangeRequestImpact `json:"impacts"`
 	ClosedStartDate *time.Time            `json:"closedStartDate"`
 	ClosedEndDate   *time.Time            `json:"closedEndDate"`
+	// Number filters to the change request whose human-readable number (e.g.
+	// "CHG0010001") exactly matches (optional). Exact match against
+	// ServiceNow's `number` column, routed as a first-class filter rather
+	// than through the free-text SearchQuery scan.
+	Number string `json:"number"`
```
[the remark this answers is not in the document]
Comment (rksk) [reply entity-service/internal/domain/entity.go]: Fixed in `e8b3abf1c` — all three `Number` fields (SearchIncidentsFilters/SearchProblemsFilters/SearchChangeRequestsFilters) are now `*string` with `omitempty`, matching the case search's own `Number`/`InternalID` (*string) exactly. Threaded through the existing `stringPtrValue` helper (sn_id.go) at each call site, same as cases already do.

File: entity-service/internal/service/sn_problem_service.go [hunk 5]
```diff
@@ -75,7 +79,7 @@ func (s *snProblemService) SearchProblems(ctx context.Context, req domain.Search
 	token := middleware.UserIDTokenFromContext(ctx)
 
 	payload := snProblemSearchPayload{
-		Filters:    snProblemFilters{SearchQuery: req.Filters.SearchQuery},
+		Filters:    snProblemFilters{SearchQuery: req.Filters.SearchQuery, Number: req.Filters.Number},
```
[the remark this answers is not in the document]
Comment (rksk) [reply entity-service/internal/service/sn_problem_service.go]: Fixed in `e8b3abf1c` — added `validateExactNumber` (sn_id.go), called before payload construction in all three of SearchProblems/SearchIncidents/SearchChangeRequests (not just problems, since the same gap existed in all three), plus a rejection test. Note: it's a length-cap only (mirrors `validateSearchQuery`'s style, 50 chars), not a format/prefix check — the case search's own `number`/`internalId` exact-match filters (case_filters.go) don't validate format either, and hardcoding a prefix/format assumption here would bake in an SN configuration detail this layer shouldn't own. Happy to tighten to a format check if you'd rather have one.

**Comment by rksk** [comment]
@coderabbitai resume

**Review by cloby99 — APPROVED** [review]
